compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.cpp

```diff
@@ -515,6 +516,73 @@ SortOp::reifyResultShapes(OpBuilder &b,
       .reifyResultShapes(b, reifiedReturnShapes);
 }
 
+namespace {
+struct RemoveUnusedSortOpResults
+    : public OpRewritePattern<IREE::LinalgExt::SortOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(IREE::LinalgExt::SortOp sortOp,
+                                PatternRewriter &rewriter) const override {
+    // To avoid problems in dispatches associated with unused results, prune
+    // them here.
+    Location loc = sortOp->getLoc();
+    auto operands = sortOp.getOutputs();
+    auto results = sortOp.getResults();
+    unsigned numRes = sortOp.getNumResults();
+
+    // If there are no results, bufferization has happened.
+    // Do not try to prune results anymore.
+    if (numRes == 0) {
+      return failure();
+    }
+
+    // If result or associated block arg is used, do not erase.
+    Block &block = sortOp.getRegion().front();
+    auto blockArgs = block.getArguments();
+    SmallVector<Value> usedBlockArgs, usedOperands, usedResults;
+    SmallVector<Type> usedResultTypes;
+    llvm::BitVector eraseArg(numRes * 2, false);
+    for (unsigned idx : llvm::seq<unsigned>(numRes)) {
+      if (!results[idx].use_empty() || !blockArgs[2 * idx].use_empty() ||
+          !blockArgs[2 * idx + 1].use_empty()) {
+        usedOperands.push_back(operands[idx]);
+        usedResults.push_back(results[idx]);
+        usedResultTypes.push_back(results[idx].getType());
+        continue;
+      }
+      eraseArg.set(2 * idx, 2 * idx + 2);
+    }
+
+    // Bail out if no pruning required.
+    if (numRes == usedOperands.size()) {
+      return failure();
+    }
+
+    // Create new op using only operands associated to used results or block
+    // args.
+    auto pruned = rewriter.create<IREE::LinalgExt::SortOp>(
+        loc, usedResultTypes,
+        /*inputs=*/ValueRange{}, usedOperands, sortOp.getDimension());
+
+    // Prune unused block arguments from copied op.
+    rewriter.inlineRegionBefore(sortOp.getRegion(), pruned.getRegion(),
+                                pruned.getRegion().begin());
+    Block &newBlock = pruned.getRegion().front();
+    for (int i = newBlock.getNumArguments(); i > 0; --i) {
+      if (eraseArg[i - 1]) {
+        newBlock.eraseArgument(i - 1);
+      }
+    }
```

Comment:
I think you can use [newBlock.eraseArguments(eraseArg)](https://github.com/llvm/llvm-project/blob/efae492ad1ba80764ec4a85f5622a8713646f970/mlir/include/mlir/IR/Block.h#L121-L123)?

Inline content of the link:

```cpp
  /// Erases the arguments that have their corresponding bit set in
  /// `eraseIndices` and removes them from the argument list.
  void eraseArguments(const BitVector &eraseIndices);
```